### PR TITLE
Si md dataloader

### PR DESCRIPTION
**Pull request by sblackburn86**
minimal dataloader for Si structures from LAMMPS. Dataloader processor assumes a folder structure for the LAMMPS run. This should be coordinated with the script to generate the data (incoming in another PR).

Resulting tensors have are structured as dict. 
TODO in a future PR: Normalization of positions  and box parameters (TBD how to handle  this)

### CODE REVIEW

File: crystal_diffusion/data/diffusion/data_preprocess.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        # get the list of runs to parse
+        assert mode in ['train', 'valid', 'test'], f"Mode should be train, valid or test. Got {mode}."
+        list_runs = [d for d in os.listdir(raw_data_dir) if os.path.isdir(os.path.join(raw_data_dir, d))
+                     and d.startswith(f"{mode}_run")]
+        list_files = []
+        for d in list_runs:
+            if f"{d}.parquet" not in os.listdir(self.data_dir):
+                df = self.parse_lammps_run(os.path.join(raw_data_dir, d))
+                if df is not None:
+                    print('hello')
```
Comment (rousseab) [inline crystal_diffusion/data/diffusion/data_preprocess.py:52]: I suggest removing this print statement.

File: tests/data/diffusion/test_data_preprocess.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    return pd.DataFrame({
+        'box': [[1, 2, 3]],
+        'x': [[60, 6, 0.6, 0.06]],
+        'y': [[120, 12, 1.2, 0.12]],
+        'z': [[180, 18, 1.8, 0.18]]
+    })
+
+
+def test_convert_coords_to_reduced(sample_coordinates):
+    # Expected output: Each coordinate divided by 10 (the box limit)
```
Comment (rousseab) [inline tests/data/diffusion/test_data_preprocess.py:104]: I think this comment is wrong? The box is [1,2,3]?

File: tests/data/diffusion/test_data_preprocess.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    for index, row in sample_coordinates.iterrows():
+        reduced_coords = LammpsProcessorForDiffusion._convert_coords_to_reduced(row)
+        expected_coords = []
+        for x, y, z in zip(row['x'], row['y'], row['z']):
+            expected_coords.extend([x, y / 2, z / 3])
```
Comment (rousseab) [inline tests/data/diffusion/test_data_preprocess.py:109]: This assumes that `box = [1,2,3]`: why not make `box` a fixture  and use it in both `sample_coordinates()` and here?

File: crystal_diffusion/data/diffusion/data_loader.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+class LammpsLoaderParameters:
+    """Base Hyper-parameters for score networks."""
+
+    batch_size: int = 64
+    num_workers: int = 0
+    max_atom: int = 64
+    spatial_dimension: int = 3  # the dimension of Euclidean space where atoms live.
+
+
+class LammpsForDiffusionDataModule(pl.LightningDataModule):  # pragma: no cover
```
Comment (rousseab) [inline crystal_diffusion/data/diffusion/data_loader.py:31]: Do we need the `#pragma: no cover`? Or is this an artifact from the cookie cutter?

File: crystal_diffusion/data/diffusion/data_loader.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+
+    @staticmethod
+    def dataset_transform(x: Dict[typing.AnyStr, typing.Any], spatial_dim: int = 3) -> Dict[str, torch.Tensor]:
+        """Format the tensors for the Datasets library.
+
+        This function is applied right after returning the objects in __getitem__ in a torch DataLoader. Everything is
+        already batched. Lightning handles the devices (gpu, cpu, mps, etc.)
+
+        Args:
+            x: raw columns from the processed data files. Should contain natom, box, type and position.
```
Comment (rousseab) [inline crystal_diffusion/data/diffusion/data_loader.py:70]: should also contain `reduced_position`.

File: crystal_diffusion/data/diffusion/data_loader.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        Returns:
+            transformed_x: formatted values as tensors
+        """
+        transformed_x = {}
+        transformed_x['natom'] = torch.as_tensor(x['natom']).long()  # resulting tensor size: (batchsize, )
+        bsize = transformed_x['natom'].size(0)
+        transformed_x['box'] = torch.as_tensor(x['box'])  # size: (batchsize, spatial dimension)
+        for pos in ['position', 'reduced_position']:
+            transformed_x[pos] = torch.as_tensor(x[pos]).view(bsize, -1, spatial_dim)
+        transformed_x['type'] = torch.as_tensor(x['type']).long()  # size: (batchsize, natom after padding)
```
Comment (rousseab) [inline crystal_diffusion/data/diffusion/data_loader.py:82]: I suggest `natom after padding` -> `max_atom`.

File: crystal_diffusion/data/diffusion/data_loader.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+
+        return transformed_x
+
+    @staticmethod
+    def pad_samples(x: Dict[typing.AnyStr, typing.Any], max_atom: int, spatial_dim: int = 3) -> Dict[str, torch.Tensor]:
+        """Pad a sample for batching.
+
+        Args:
+            x: initial sample from the dataset. Should contain natom, position, and type.
```
Comment (rousseab) [inline crystal_diffusion/data/diffusion/data_loader.py:91]: should also contain reduced positions.

File: crystal_diffusion/data/diffusion/data_loader.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+                                                               cache_dir=self.working_cache_dir)
+            # TODO QoL valid dataset is labelled as train split by Datasets. Find a way to rename.
+        else:
+            raise NotImplementedError("Test mode needs to be implemented.")
+        # TODO test dataset when stage == 'test'
+        # we can filter out samples at this stage using the .filter(lambda x: f(x)) with f(x) a boolean function
+        # or a .select(list[int]) with a list of indices to keep a subset. This is much faster than .filter
+        # padding needs to be done here OR in the preprocessor
+        # check if the max number of atoms matches at least the max in the training set
+        self.max_atom = max(max(self.train_dataset['natom']), self.max_atom)
```
Comment (rousseab) [inline crystal_diffusion/data/diffusion/data_loader.py:125]: Here you are potentially changing an input parameter... I would prefer an assert that forces the user to set a correct value. Alternatively, send a warning through the `Logger`?

File: crystal_diffusion/data/diffusion/data_loader.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+        # map() are applied once, not in-place.
+        # The kwarg batch can accelerate by working with batches, not useful for padding
```
Comment (rousseab) [inline crystal_diffusion/data/diffusion/data_loader.py:127]: Do you mean "the keyword argument 'batched'"?

File: crystal_diffusion/data/diffusion/data_loader.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+        return DataLoader(
+            self.valid_dataset,
+            batch_size=self.batch_size,
+            shuffle=False,
+            num_workers=self.num_workers,
+        )
+
+    def test_dataloader(self):
+        """Creates the testing dataloader using the testing data parser."""
+        return DataLoader(
```
Comment (rousseab) [inline crystal_diffusion/data/diffusion/data_loader.py:157]: I suggest we raise an `NotImplementedError` here until we get to it...

**Review by rousseab — CHANGES_REQUESTED** [review]
I make a few suggestions, mostly correcting docstrings and comments.

**Review by rousseab — APPROVED** [review]
LGTM!
